dropgz/pkg/embed/payload.go

```diff
@@ -0,0 +1,110 @@
+package embed
+
+import (
+	"bufio"
+	"compress/gzip"
+	"embed"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	cwd        = "gz"
+	pathPrefix = cwd + string(filepath.Separator)
+)
+
+var ErrArgsMismatched = errors.New("mismatched argument count")
+
+// gzfs contains the gzipped files for deployment, as a read-only FileSystem containing only "gzfs/".
+//nolint:typecheck // dir is populated at build.
+//go:embed gz
+var gzfs embed.FS
+
+func Contents() ([]string, error) {
+	contents := []string{}
+	if err := fs.WalkDir(gzfs, cwd, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		contents = append(contents, strings.TrimPrefix(path, pathPrefix))
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "error walking gzfs")
+	}
+	return contents, nil
+}
+
+// gzipCompoundReadCloser is a wrapper around the source file handle and
+// the gzip Reader on the file to provide a single Close implementation
+// which cleans up both.
+// We have to explicitly track and close the underlying Reader, because
+// the gzip readercloser# does not.
+type gzipCompoundReadCloser struct {
+	file     io.Closer
+	gzreader io.ReadCloser
+}
+
+func (rc *gzipCompoundReadCloser) Read(p []byte) (n int, err error) {
+	return rc.gzreader.Read(p)
+}
+
+func (rc *gzipCompoundReadCloser) Close() error {
+	if err := rc.gzreader.Close(); err != nil {
+		return err
+	}
+	if err := rc.file.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func Extract(path string) (*gzipCompoundReadCloser, error) {
+	f, err := gzfs.Open(filepath.Join(cwd, path))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open file %s", path)
+	}
+	r, err := gzip.NewReader(bufio.NewReader(f))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to build gzip reader")
+	}
+	return &gzipCompoundReadCloser{file: f, gzreader: r}, nil
+}
+
+func deploy(src, dest string) error {
+	rc, err := Extract(src)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	target, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o766) //nolint:gomnd // executable file bitmask
```

Comment:
I understand the immediate intention is deploying executables, but it might make sense to add the file's mode to `gzipCompoundReadCloser` so that you can recreate files with the same mode here. It might get... um, interesting with things like symlink, directory, and pipe mode bits. Though, at least if you wanted to do SUID or have a read-only config, that should work too.

Reply:
copying the file mode from the embed is a much better idea 🤦🏼

Reply:
it's not feasible to keep the file perms without introducing additional dependencies (tar) or tabulating and restoring the bitmask ourselves. i think it is reasonable to deploy 766 and say that the caller can chmod if necessary.